PurchasesCoreSwift/Purchasing/PurchasesOrchestrator.swift

```diff
@@ -0,0 +1,506 @@
+
+                guard let keyIdentifier = maybeKeyIdentifier,
+                      let nonce = maybeNonce,
+                      let signature = maybeSignature,
+                      let timestamp = maybeTimestamp else {
+                          // TODO: Consider making a custom error.
+                          completion(nil, ErrorUtils.unexpectedBackendResponseError())
+                          return
+                      }
+
+                let paymentDiscount = SKPaymentDiscount(identifier: discountIdentifier,
+                                                        keyIdentifier: keyIdentifier,
+                                                        nonce: nonce,
+                                                        signature: signature,
+                                                        timestamp: timestamp)
+                completion(paymentDiscount, nil)
+            }
+        }
+    }
+
+    @objc public func purchase(product: SKProduct,
+                               payment: SKMutablePayment,
+                               presentedOfferingIdentifier maybePresentedOfferingIdentifier: String?,
+                               completion: @escaping PurchaseCompletedBlock) {
+        Logger.debug("makePurchase")
+        guard let productIdentifier = extractProductIdentifier(fromProduct: product, orPayment: payment) else {
+            Logger.info(Strings.purchase.could_not_purchase_product_id_not_found)
+            let errorMessage = "There was a problem purchasing the product: productIdentifier was nil"
+            completion(nil, nil, ErrorUtils.unknownError(message: errorMessage), false)
+            return
+        }
+
+        if !self.finishTransactions {
+            Logger.warn(Strings.purchase.purchasing_with_observer_mode_and_finish_transactions_false_warning)
+        }
+
+        let currentAppUserID = self.appUserID
+        payment.applicationUsername = currentAppUserID
+
+        preventPurchasePopupCallFromTriggeringCacheRefresh(appUserID: appUserID)
+
+        if let presentedOfferingIdentifier = maybePresentedOfferingIdentifier {
+            Logger.purchase(String(format: Strings.purchase.purchasing_product_from_package,
+                                   productIdentifier,
+                                   presentedOfferingIdentifier))
+            lock.lock()
+            presentedOfferingIDsByProductID[productIdentifier] = presentedOfferingIdentifier
+            lock.unlock()
+
+        } else {
+            Logger.purchase(String(format: Strings.purchase.purchasing_product, productIdentifier))
+        }
+
+        productsManager.cacheProduct(product)
+
+        lock.lock()
+        guard purchaseCompleteCallbacksByProductID[productIdentifier] == nil else {
+            completion(nil, nil, ErrorUtils.operationAlreadyInProgressError(), false)
+            lock.unlock()
+            return
+        }
+        purchaseCompleteCallbacksByProductID[productIdentifier] = completion
```

Comment:
will we unlock if we don't go into that guard?

Reply:
🤦‍♂️  good catch, it won't!